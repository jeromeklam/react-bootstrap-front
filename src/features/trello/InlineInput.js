import React from 'react';
import PropTypes from 'prop-types';
import autosize from 'autosize';

class InlineInput extends React.Component {

  static propTypes = {
    onSave: PropTypes.func,
    border: PropTypes.bool,
    placeholder: PropTypes.string,
    value: PropTypes.string,
    autoFocus: PropTypes.bool,
    resize: PropTypes.oneOf(['none', 'vertical', 'horizontal']),
  };

  static defaultProps = {
    onSave: () => {},
    placeholder: '',
    value: '',
    border: false,
    autoFocus: false,
    resize: 'none',
  };

  static getDerivedStateFromProps(props, state) {
    if (props.value !== state.value) {
      return { value: props.value };
    }
    return null;
  }

  constructor(props) {
    super(props);
    this.state = {
      value: props.value,
    }
  }

  onFocus = e => e.target.select();

  // This is the way to select all text if mouse clicked
  onMouseDown = e => {
    if (document.activeElement !== e.target) {
      e.preventDefault();
      this.refInput.focus();
    }
  };

  onBlur = () => {
    this.updateValue();
  };

  onKeyDown = e => {
    if (e.keyCode === 13) {
      this.refInput.blur();
      e.preventDefault();
    }
    if (e.keyCode === 27) {
      this.setValue(this.props.value);
      this.refInput.blur();
      e.preventDefault();
    }
    if (e.keyCode === 9) {
      if (this.getValue().length === 0) {
        this.props.onCancel();
      }
      this.refInput.blur();
      e.preventDefault();
    }
  };

  getValue = () => this.refInput.value;
  setValue = value => (this.refInput.value = value);

  updateValue = () => {
    if (this.getValue() !== this.props.value) {
      this.props.onSave(this.getValue());
    }
  };

  setRef = ref => {
    this.refInput = ref;
    if (this.props.resize !== 'none') {
      autosize(this.refInput);
    }
  };

  render() {
    const { autoFocus, border, value, placeholder } = this.props;
    return (
      <textarea
        className="trello-inline-input text-secondary bold"
        ref={this.setRef}
        onMouseDown={this.onMouseDown}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
        onKeyDown={this.onKeyDown}
        placeholder={value.length === 0 ? undefined : placeholder}
        defaultValue={value}
        autoComplete="off"
        autoCorrect="off"
        autoCapitalize="off"
        spellCheck="false"
        rows={1}
        autoFocus={autoFocus}
      />
    );
  }
};

export default InlineInput;
