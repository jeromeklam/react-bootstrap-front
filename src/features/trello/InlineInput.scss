.trello-inline-input {
  font-weight: bold;
  &:not(:focus) {
    background-color: transparent;
    border: theme-color('light');
  }
}