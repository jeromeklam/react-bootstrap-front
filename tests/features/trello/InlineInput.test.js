import React from 'react';
import renderer from 'react-test-renderer';
import InlineInput from '../../../src/features/trello/InlineInput';

const callback = jest.fn();
jest.useFakeTimers();

it('render with className and name', () => {
  // First render
  const component = renderer.create(
    <InlineInput className="btn bg-secondary" onClick={callback}>
      My Button
    </InlineInput>
  );
  let test = component.toJSON();
  expect(test).toMatchSnapshot();
});
