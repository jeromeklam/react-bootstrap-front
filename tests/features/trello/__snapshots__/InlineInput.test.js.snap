// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render with className and name 1`] = `
<textarea
  autoCapitalize="off"
  autoComplete="off"
  autoCorrect="off"
  autoFocus={false}
  border={false}
  className="trello-inline-input"
  dataGramm="false"
  defaultValue=""
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  rows={1}
  spellCheck="false"
/>
`;
